File: services/agent-service/src/intelligence/merkle.ts
import { createHash } from 'node:crypto';

/**
 * arcana-anchor/v1 — the same construction as
 * services/decision-engine/internal/store/merkle.go, which BUILDS the roots.
 * This one CHECKS them. Two implementations on purpose: a proof checked by the
 * code that produced it proves only that the code agrees with itself.
 *
 *   leaf  = sha256(0x00 || commitment)      commitment as its 32 raw bytes
 *   node  = sha256(0x01 || left || right)
 *   odd   = the last node of a level is carried up unchanged
 *   order = decision id ascending
 *
 * On chain: a zero-value transaction from the anchoring address to itself whose
 * input is ANCHOR_MAGIC_HEX followed by the root.
 */
export const ANCHOR_SCHEME = 'arcana-anchor/v1';
/** "ARCANA" 0x00 0x01. */
export const ANCHOR_MAGIC_HEX = '415243414e410001';

const sha256 = (...parts: Buffer[]) => {
  const h = createHash('sha256');
  for (const p of parts) h.update(p);
  return h.digest();
};

export function leafHash(commitment: string): Buffer {
  const c = commitment.trim().toLowerCase();
  if (!/^[0-9a-f]{64}$/.test(c)) throw new Error(`commitment ${commitment} is not 32 bytes of hex`);
  return sha256(Buffer.from([0x00]), Buffer.from(c, 'hex'));
}

const node = (left: Buffer, right: Buffer) => sha256(Buffer.from([0x01]), left, right);

function nextLevel(level: Buffer[]): Buffer[] {
  const out: Buffer[] = [];
  for (let i = 0; i < level.length; i += 2) {
    out.push(i + 1 === level.length ? level[i] : node(level[i], level[i + 1]));
  }
  return out;
}

export function merkleRoot(leaves: Buffer[]): Buffer | null {
  if (leaves.length === 0) return null;
  let level = leaves;
  while (level.length > 1) level = nextLevel(level);
  return level[0];
}

export type ProofStep = { sibling: string; position: 'left' | 'right' };

export function merkleProof(leaves: Buffer[], index: number): ProofStep[] {
  if (index < 0 || index >= leaves.length) throw new Error(`leaf ${index} outside ${leaves.length} leaves`);
  const proof: ProofStep[] = [];
  let level = leaves;
  let i = index;
  while (level.length > 1) {
    if (i % 2 === 1) proof.push({ sibling: level[i - 1].toString('hex'), position: 'left' });
    else if (i + 1 < level.length) proof.push({ sibling: level[i + 1].toString('hex'), position: 'right' });
    level = nextLevel(level);
    i = Math.floor(i / 2);
  }
  return proof;
}

export function verifyProof(leaf: Buffer, proof: ProofStep[], root: Buffer): boolean {
  let cur = leaf;
  for (const step of proof) {
    const sib = Buffer.from(step.sibling, 'hex');
    cur = step.position === 'left' ? node(sib, cur) : node(cur, sib);
  }
  return cur.equals(root);
}

File: services/decision-engine/internal/store/merkle.go
package store

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
)

// AnchorScheme names the Merkle construction and the on-chain payload. See
// migration 0048 and docs/anchoring.md.
//
//	leaf  = sha256(0x00 || commitment)      commitment as its 32 raw bytes
//	node  = sha256(0x01 || left || right)
//	odd   = the last node of a level is carried up unchanged
//	order = decision id ascending
//
// DOMAIN SEPARATION is the 0x00/0x01 prefix. Without it, an interior node is
// 32 bytes like any leaf, and a proof could present a node as if it were a
// commitment. CARRYING the odd node rather than pairing it with itself is the
// other deliberate choice: duplicating the last leaf makes two different leaf
// lists share a root, which is a known way to forge a Merkle membership.
const AnchorScheme = "arcana-anchor/v1"

// LeafHash hashes one decision commitment (64 hex characters) into a leaf.
func LeafHash(commitment string) ([]byte, error) {
	b, err := hex.DecodeString(strings.TrimSpace(commitment))
	if err != nil || len(b) != 32 {
		return nil, fmt.Errorf("commitment %q is not 32 bytes of hex", commitment)
	}
	h := sha256.New()
	h.Write([]byte{0x00})
	h.Write(b)
	return h.Sum(nil), nil
}

func nodeHash(left, right []byte) []byte {
	h := sha256.New()
	h.Write([]byte{0x01})
	h.Write(left)
	h.Write(right)
	return h.Sum(nil)
}

// MerkleRoot returns the root over leaf hashes, in the order given. Nil for no
// leaves: there is no root of nothing, and anchoring one would prove nothing.
func MerkleRoot(leaves [][]byte) []byte {
	if len(leaves) == 0 {
		return nil
	}
	level := leaves
	for len(level) > 1 {
		level = nextLevel(level)
	}
	return level[0]
}

func nextLevel(level [][]byte) [][]byte {
	next := make([][]byte, 0, (len(level)+1)/2)
	for i := 0; i < len(level); i += 2 {
		if i+1 == len(level) {
			next = append(next, level[i])
			continue
		}
		next = append(next, nodeHash(level[i], level[i+1]))
	}
	return next
}

// ProofStep is one sibling on the path from a leaf to the root.
type ProofStep struct {
	Sibling       []byte
	SiblingOnLeft bool
}

// MerkleProof returns the path for leaves[index]. A level where the node is the
// carried odd one contributes no step.
func MerkleProof(leaves [][]byte, index int) ([]ProofStep, error) {
	if index < 0 || index >= len(leaves) {
		return nil, fmt.Errorf("leaf index %d outside %d leaves", index, len(leaves))
	}
	var proof []ProofStep
	level, i := leaves, index
	for len(level) > 1 {
		if i%2 == 1 {
			proof = append(proof, ProofStep{Sibling: level[i-1], SiblingOnLeft: true})
		} else if i+1 < len(level) {
			proof = append(proof, ProofStep{Sibling: level[i+1], SiblingOnLeft: false})
		}
		level = nextLevel(level)
		i /= 2
	}
	return proof, nil
}

// VerifyProof walks a proof from a leaf and reports whether it reaches root.
func VerifyProof(leaf []byte, proof []ProofStep, root []byte) bool {
	cur := leaf
	for _, s := range proof {
		if s.SiblingOnLeft {
			cur = nodeHash(s.Sibling, cur)
		} else {
			cur = nodeHash(cur, s.Sibling)
		}
	}
	return bytes.Equal(cur, root)
}
